tools/prngd-0.9.29/prngd.man
.\" Automatically generated by Pod::Man version 1.15
.\" Tue Jul 13 09:11:19 2004
.\"
.\" Standard preamble:
.\" ======================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Ip \" List item
.br
.ie \\n(.$>=3 .ne \\$3
.el .ne 3
.IP "\\$1" \\$2
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R

.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  | will give a
.\" real vertical bar.  \*(C+ will give a nicer C++.  Capital omega is used
.\" to do unbreakable dashes and therefore won't be available.  \*(C` and
.\" \*(C' expand to `' in nroff, nothing in troff, for use with C<>
.tr \(*W-|\(bv\*(Tr
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr
.\" for titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and
.\" index entries marked with X<> in POD.  Of course, you'll have to process
.\" the output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it
.\" makes way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.bd B 3
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ======================================================================
.\"
.IX Title "PRNGD 1"
.TH PRNGD 1 "" "2002-05-17" ""
.UC
.SH "NAME"
prngd \- Pseudo Random Number Generator Daemon
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBprngd\fR
[ \fIcommand_opts\fR ]
\&\fI/path/to/socket\fR
[ \fI/path/to/other_socket\fR ]
.PP
\&\fBprngd\fR
\&\fI\-k/\-\-killmode\fR
\&\fI/path/to/socket\fR
[ \fI/path/to/other_socket\fR ]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\s-1PRNGD\s0 is the Pseudo Random Number Generator Daemon.
It offers an \s-1EGD\s0 compatible interface to obtain random data and is
intended to be used as an entropy source to feed other software,
especially software based on OpenSSL.
.PP
Like \s-1EGD\s0 (http://sourceforge.net/projects/egd/) it calls system programs
to collect entropy.  Unlike \s-1EGD\s0 it does not generate a pool of random
bits that can be called from other software.
.PP
Rather it feeds the bits gathered into its internal \s-1PRNG\s0 from which
the \*(L"random bits\*(R" are obtained when requested. This way, \s-1PRNGD\s0 is
never drained and can never block (unlike \s-1EGD\s0), so it is also suitable
to seed inetd-started programs.
.PP
It also features a seed-save file, so that it is immediately usable
after system start.
.SH "USAGE"
.IX Header "USAGE"
Startup \- (reads the seed file)
 /usr/local/bin/prngd [other options] /var/run/egd-pool
.PP
Clean shutdown \- (saves current data to the seed file)
 /usr/local/bin/prngd \-\-kill /var/run/egd-pool
.PP
The above line should be added to the system startup files so that
prngd starts up before any program that would query the random data
socket, such as OpenSSH.
.PP
\&\s-1PRNGD\s0 supports both \s-1UNIX\s0 sockets (as shown in the examples) as well as
\&\s-1TCP\s0 sockets (localhost only!). There is no port number registered (yet),
so pick a free port on your system, e.g. 708 when running with root perms
and you want to offer it as a system service, or 4840 without root perms.
.PP
Both ports are not assigned to any service as of
 http://www.isi.edu/in-notes/iana/assignments/port-numbers
.PP
To have \s-1PRNGD\s0 listen on both the Unix socket and \s-1TCP\s0 do
 prngd [other options] /var/run/egd-pool tcp/localhost:708 tcp/localhost:4840
.PP
You should add these lines to /etc/services: (substitute your chosen ports)
 prngd         708/tcp     # prngd/EGD system service
 prngd-user    4840/tcp    # prngd/EGD user service
.PP
\&\s-1PRNGD\s0 itself by now only handles numerical port numbers.
.SH "OPTIONS"
.IX Header "OPTIONS"
\&\s-1PRNGD\s0 has the following options:
.Ip "\-c / \-\-cmdfile file" 4
.IX Item "-c / --cmdfile file"
Read the entropy gathering commands from \*(L"file\*(R" instead of reading from
the compiled in default location.
.Ip "\-d / \-\-debug" 4
.IX Item "-d / --debug"
Do not go into daemon (background) mode and print diagnostic messages to
stderr.
.Ip "\-f / \-\-fg" 4
.IX Item "-f / --fg"
Stay in foreground, do not fork. Do not print diagnostic messages. This
is e.g. usefull when using \s-1AIX\s0 System Resource Controller, so that
the resource controler could detect a crash of prngd.
.Ip "\-k / \-\-kill" 4
.IX Item "-k / --kill"
Kill the \fIdaemon\fR\|(s) at the locations given. It is possible to kill more than
one prngd at a time with this command. If several locations are given at
one time, prngd will try to contact them one after the other, query the
process id according to the \s-1EGD\s0 protocol and send a \s-1SIGTERM\s0 to the
process. As one daemon may server more than one socket, one may wish
to kill the daemon with the same argument that was used to start prngd.
The kill operation will however only succeed for the first socket, as
the daemon will be killed in the first attempt. prngd \-k ... will therefore
silently ignore the failure at the other \fIlocation\fR\|(s).
.Ip "\-m / \-\-mode mode" 4
.IX Item "-m / --mode mode"
Set the file access mode of the \s-1UNIX\s0 socket to be \*(L"mode\*(R". The mode is
set after creation of the socket. This option can be used to restrict
access to prngd. If more than one \s-1UNIX\s0 socket is served by prngd, the
same mode setting applies to all \s-1UNIX\s0 sockets. The mode must be given
in numeric notation, e.g. 777 for read and write and execute permission
for owner/group/other.
.Ip "\-n / \-\-no-seedfile" 4
.IX Item "-n / --no-seedfile"
Do not read from any seedfile and do not create a seedfile or write to
any seedfile.
.Ip "\-p / \-\-pidfile file" 4
.IX Item "-p / --pidfile file"
Write the pid of the prngd process into the specified file. If this option
is not explicitly used, no pid-file will be written, as prngd \-k can be
used to kill a running prngd.
.Ip "\-s / \-\-seedfile file" 4
.IX Item "-s / --seedfile file"
Location of the seedfile. The file is read at startup to initially
seed the \s-1PRNG\s0. On clean shutdown entropy is retrieved from the pool
and written back to the file. If no location is given, the compiled in
location of the seedfile is used. If the seedfile does not exist on startup,
it is created automatically.
.Ip "\-v / \-\-version" 4
.IX Item "-v / --version"
Print the version of prngd and exit.
.SH "DIAGNOSTICS"
.IX Header "DIAGNOSTICS"
.SH "AUTHOR/LICENSE"
.IX Header "AUTHOR/LICENSE"
Author: Lutz Jaenicke <Lutz.Jaenicke@aet.TU-Cottbus.DE>
.PP
This software is free. You can do with it whatever you want.
I would however kindly ask you to acknowledge the use of this
package, if you are going use it in your software, which you might
be going to distribute. I would also like to receive a note if you
are a satisfied user :\-)
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\s-1PRNGD\s0 Home Page:
  http://www.aet.tu-cottbus.de/personen/jaenicke/postfix_tls/prngd.html
.PP
Related Sites:
  http://sourceforge.net/projects/egd/
  http://www.openssh.org/
  http://www.openssl.org/


CHANGES
update nginx to 1.11.8
add tools:
  prngd-0.9.29
add to src libs:
  zlib-1.2.8 (fix for windows with unistd.h)
  pcre-8.39
add submodule openssl
change branch openssl to OpenSSL_1_1_0-stable
add modules:
  nginx-let-module
add builds for QNX


tools/prngd-0.9.29/prngd.c
#include <stdlib.h>
#include <stdio.h>
#include <errno.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <signal.h>
#include <time.h>
#include <assert.h>
#include <syslog.h>
#include <pwd.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/time.h>
#include <sys/wait.h>

#include "prngd.h"

static int randsavefd = -1;
static int pidfilefd = -1;
char *pidfile = NULL;
int debug = 0;
int fg = 0;

int devnull = -1;

int limit_openfd = 0;
int count_openfd = 0;

static void write_seedfile(void);
static void open_pidfile(const char *pidfilename)
{
  char msg_buf[1024];

  remove(pidfilename);
  pidfilefd = open(pidfilename, O_WRONLY | O_CREAT, 400);
  if (pidfilefd < 0)
  {
    (void)snprintf(msg_buf, sizeof(msg_buf),
	"could not open %s: %s\n", pidfilename, strerror(errno));
    msg_buf[sizeof(msg_buf) - 1] = '\0';
    error_fatal(msg_buf, 0);
    exit(EXIT_FAILURE);
  }
  close_on_exec(pidfilefd, 1);
}


static void open_seedfile(const char *seedfile)
{
  int bytes, bytes_read_total;
  unsigned char buffer[PRNGD_STATE_SIZE];
  char msg_buf[1024];

  randsavefd = open(seedfile, O_RDWR | O_CREAT | O_SYNC | O_RSYNC, 400);
  if (randsavefd < 0)
  {
    (void)snprintf(msg_buf, sizeof(msg_buf),
	"could not open %s: %s\n", seedfile, strerror(errno));
    msg_buf[sizeof(msg_buf) - 1] = '\0';
    error_fatal(msg_buf, 0);
    exit(EXIT_FAILURE);
  }
  close_on_exec(randsavefd, 1);

  /*
   * The contents of the seedfile may be sensitive, since we start seeding
   * from the file.
   */
  if (chmod(seedfile, S_IRUSR | S_IWUSR) < 0)
  {
    (void)snprintf(msg_buf, sizeof(msg_buf),
	"Could not chmod %s: %s\n", seedfile, strerror(errno));
    msg_buf[sizeof(msg_buf) - 1] = '\0';
    error_fatal(msg_buf, 0);
    exit(EXIT_FAILURE);
  }

  /*
   * Do some initial seeding stuff to pre-stir the PRNG.
   */
  seed_internal(NULL);

  /*
   * Now read as much from the file as we can get. 
   */
  bytes_read_total = 0;
  while ((bytes = read(randsavefd, buffer, PRNGD_STATE_SIZE)) > 0)
  {
    rand_add(buffer, bytes, 0.0);
    bytes_read_total += bytes;
  }
  if (bytes < 0)
  {
    (void)snprintf(msg_buf, sizeof(msg_buf),
	"Error reading %s: %s\n", seedfile, strerror(errno));
    msg_buf[sizeof(msg_buf) - 1] = '\0';
    error_fatal(msg_buf, 0);
    exit(EXIT_FAILURE);
  }

  /*
   * Done, let's stir around some more
   */
  seed_internal(NULL);

  if (debug)
    (void)fprintf(stderr, "Read %d bytes\n", bytes_read_total);
  /*
   * Finally, account for the bytes read. We don't know how much entropy
   * was in the bytes we've read, so we only account for the minimum needed
   * and have PRNGD do the rest. If not enough bytes are read, only account
   * the bytes. Don't care for the contents of the buffer, it doesn't hurt
   * to stir the pool.
   */
  if (bytes_read_total > ENTROPY_NEEDED)
    rand_add(buffer, PRNGD_STATE_SIZE, ENTROPY_NEEDED);
  else
    rand_add(buffer, PRNGD_STATE_SIZE, bytes_read_total);

  /*
   * Ok, we don't want to reuse the same seed again, so we will immediately
   * replace the contents of the seed-file.
   */
  write_seedfile();
}

static void write_seedfile(void)
{
  int bytes;
  unsigned char buffer[PRNGD_STATE_SIZE];

  bytes = lseek(randsavefd, 0, SEEK_SET);
  if (bytes < 0)
  {
    error_fatal("Error writing seedfile", errno);
    exit(EXIT_FAILURE);
  }
  bytes = rand_bytes(buffer, PRNGD_STATE_SIZE);
  if (bytes == 0)
  {
    error_error("Info: Random pool not (yet) seeded", 0);
  }
  else
  {
    bytes = write(randsavefd, buffer, PRNGD_STATE_SIZE);
    if (bytes < 0)
    {
      error_fatal("Error writing seedfile", errno);
      exit(EXIT_FAILURE);
    }
  }
  if (debug)
    (void)fprintf(stderr, "Wrote %d bytes back to seedfile\n", bytes);

  bytes = ftruncate(randsavefd, bytes);
}

void close_seedfile(void)
{
  if (randsavefd >= 0) {
    write_seedfile();
    (void)close(randsavefd);
  }
}

static void close_pidfile(void)
{
  if (pidfilefd >= 0) {
    (void)close(pidfilefd);
  }
}


static void killpeer(int service_socket)
{
  unsigned char buffer[256];
  char msg_buf[1024];
  int num;
  pid_t peer_id;

  buffer[0] = 0x04;
  if (write(service_socket, buffer, 1) < 1)
  {
    error_fatal("Cannot contact peer", errno);
    exit(EXIT_FAILURE);
  }
  if (read(service_socket, buffer, 1) < 1)
  {
    error_fatal("Cannot read from peer", errno);
    exit(EXIT_FAILURE);
  }
  num = buffer[0];
  if (read(service_socket, buffer, num) < num)
  {
    error_fatal("Cannot read from peer", errno);
    exit(EXIT_FAILURE);
  }
  buffer[num] = '\0';
  peer_id = (pid_t)atol((char *) buffer);
  (void)fprintf(stderr, "Killing: %ld\n", (long)peer_id);

  if (kill(peer_id, SIGTERM) < 0)
  {
    (void)snprintf(msg_buf, sizeof(msg_buf), "Cannot kill %ld", (long)peer_id);
    msg_buf[sizeof(msg_buf) - 1] = '\0';
    error_fatal(msg_buf, errno);
    exit(EXIT_FAILURE);
  }
  exit(EXIT_SUCCESS);
}

static void usage(const char *progname)
{
  (void)fprintf(stderr, "prngd %s (%s)\n", PRNGD_VERSION, PRNGD_DATE);
  (void)fprintf(stderr,
	  "Usage: %s [options] (/path/to/socket1 | tcp/localhost:port) "
		"[(/path/to/socket2 | tcp/localhost:port)] ... \n"
	  "Options:\n"
	  "\t-f/--fg: do not fork\n"
	  "\t-d/--debug: debugging on\n"
	  "\t-c/--cmdfile cmdpath: use cmdpath for entropy commands [%s]\n"
	  "\t-s/--seedfile seedpath: use seedpath as seedfile [%s]\n"
	  "\t-p/--pidfile pidfilepath: use pidfilepath as pidfile\n"
	  "\t-n/--no-seedfile: no seedfile, keep pool in memory only\n"
	  "\t-m/--mode mode: use mode for sockets [%04o]\n"
	  "\t-k/--kill: kill daemon on other side\n"
	  "\t-v/--version: print version and exit\n",
	  progname, CONFIGFILE, RANDSAVENAME, SOCKETMODE);
}

int main(int argc, char *argv[])
{
  int argp;
  int ret;
  int statval;
  int killmode = 0, killfail = 0;
  time_t seed_stat_interval = SEED_STAT_INTERVAL;
  time_t seed_ext_interval = SEED_EXT_INTERVAL;
  int i, numsock_file = 0, numsock_tcp = 0;
  char **randsock_tcp = NULL;
  char **randsock_file = NULL;
  char *cmdpath = CONFIGFILE;
  char *seedpath = RANDSAVENAME;
  char msg_buf[1024];
  unsigned int socketmode = SOCKETMODE;
  int *service_socket, service_count = 0;
  int kill_socket;
  struct stat randsock_file_stat;
  struct passwd *pwent;
  char *username = "unknown";
  char pidfilebuffer[32];
  entropy_source_t *entropy_source;

  /*
   * Process arguments 
   */
  argp = 1;
  while (argp < argc)
  {
    if (!strcmp(argv[argp], "-v") || !strcmp(argv[argp], "--version")) {
      (void)fprintf(stdout, "prngd %s (%s)\n", PRNGD_VERSION, PRNGD_DATE);
      exit(EXIT_SUCCESS);
    }
    else if (!strcmp(argv[argp], "-d") || !strcmp(argv[argp], "--debug"))
      debug = 1;
    else if (!strcmp(argv[argp], "-f") || !strcmp(argv[argp], "--fg"))
      fg = 1;
    else if ((!strcmp(argv[argp], "-c") || !strcmp(argv[argp], "--cmdfile"))
	     && (argp + 1 < argc))
      cmdpath = argv[++argp];
    else if ((!strcmp(argv[argp], "-s") || !strcmp(argv[argp], "--seedfile"))
	     && (argp + 1 < argc))
      seedpath = argv[++argp];
    else if ((!strcmp(argv[argp], "-p") || !strcmp(argv[argp], "--pidfile"))
	     && (argp + 1 < argc))
      pidfile = argv[++argp];
    else if ((!strcmp(argv[argp], "-m") || !strcmp(argv[argp], "--mode"))
	     && (argp + 1 < argc))
    {
      if (sscanf(argv[++argp], "%o", &socketmode) == 0)
      {
	(void)snprintf(msg_buf, sizeof(msg_buf),
		"Cannot interpret socket mode '%s'\n", argv[argp]);
	msg_buf[sizeof(msg_buf) - 1] = '\0';
	error_fatal(msg_buf, 0);
	exit(EXIT_FAILURE);
      }
    }
    else if (!strcmp(argv[argp], "-n") || !strcmp(argv[argp], "--no-seedfile"))
      seedpath = NULL;
    else if (!strcmp(argv[argp], "-k") || !strcmp(argv[argp], "--kill"))
      killmode = 1;
#ifndef NO_UNIX_SOCKET
    else if (argv[argp][0] == '/') {
      if (numsock_file == 0)
        randsock_file = malloc(sizeof(char *));
      else
	randsock_file = realloc(randsock_file,
				(numsock_file + 1) * sizeof(char *));
      if (randsock_file == NULL) {
	error_fatal("Could not allocate memory", errno);
	exit(EXIT_FAILURE);
      }
      randsock_file[numsock_file++] = argv[argp];
    }
#endif /* NO_UNIX_SOCKET */
    else if (!strncmp(argv[argp], "tcp/localhost:", strlen("tcp/localhost:"))) {
      if (numsock_tcp == 0)
        randsock_tcp = malloc(sizeof(char *));
      else
	randsock_tcp = realloc(randsock_tcp,
				(numsock_tcp + 1) * sizeof(char *));
      if (randsock_tcp == NULL) {
	error_fatal("Could not allocate memory", errno);
	exit(EXIT_FAILURE);
      }
      randsock_tcp[numsock_tcp++] = argv[argp];
    } 
    else
    {
      usage(argv[0]);
      exit(EXIT_FAILURE);
    }
    argp++;
  }
  if ((numsock_file == 0) && (numsock_tcp == 0)) {
    usage(argv[0]);
    error_fatal("No listen socket given, exiting!", 0);
    exit(EXIT_FAILURE);
  }

  if (!killmode)
    parse_configfile(cmdpath, &entropy_source);

  if (debug)
    (void)fprintf(stdout, "Debugging enabled\n");

  /*
   * Obtain the maximum number of open filedescriptors and adjust the
   * counter of actual open files. As always we have open stdin, stdout,
   * and stderr.
   */
  limit_openfd = obtain_limit();
  count_openfd = 3;

  /*
   * Open /dev/null
   */
  devnull = open("/dev/null", O_RDWR);
  if (devnull == -1) {
    error_fatal("Couldn't open /dev/null", errno);
    exit(EXIT_FAILURE);
  }
  count_openfd++;

  /*
   * Now, initialize the PRNG (allocate the memory).
   */
  rand_init(PRNGD_STATE_SIZE, ENTROPY_NEEDED);


  /*
   * Start by reading back the seed saved from former runs, if usage of a
   * seedfile is wanted. 
   */
  if (!killmode && seedpath) {
    open_seedfile(seedpath);
    count_openfd++;
  }
  if (!killmode && pidfile) {
    open_pidfile(pidfile);
    count_openfd++;
  }

  if (!(service_socket = malloc((numsock_file + numsock_tcp) * sizeof(int)))) {
    error_fatal("Could not allocate memory", errno);
    exit(EXIT_FAILURE);
  }

  if (killmode) {
    for (i = 0; i < numsock_file; i++) {

    /*
     * Try to kill daemons on the other side of a unix socket. If we cannot
     * connect to a daemon, we set a flag but continue, since there may be
     * other daemons waiting for kill.
     */
      statval = stat(randsock_file[i], &randsock_file_stat);
      if ((statval < 0) && (errno != ENOENT)) {
        (void)snprintf(msg_buf, sizeof(msg_buf),
		       "Cannot stat socket position %s", randsock_file[i]);
        msg_buf[sizeof(msg_buf) - 1] = '\0';
        error_error(msg_buf, errno);
        killfail = 1;
        continue;
      }
      if ((statval < 0) && (errno == ENOENT)) {
        (void)snprintf(msg_buf, sizeof(msg_buf), "No object at position %s",
        	randsock_file[i]);
        msg_buf[sizeof(msg_buf) - 1] = '\0';
        error_fatal(msg_buf, errno);
        killfail = 1;
        continue;
      }
      if (!IS_UNIX_SOCKET(randsock_file_stat.st_mode)) {
	(void)snprintf(msg_buf, sizeof(msg_buf), "Will not touch %s: no socket",
		randsock_file[i]);
	msg_buf[sizeof(msg_buf) - 1] = '\0';
	error_error(msg_buf, errno);
	killfail = 1;
	continue;
      }

      if ((kill_socket = unix_connect(randsock_file[i], 0, 1)) < 0) {
	    (void)snprintf(msg_buf, sizeof(msg_buf),
	    	"cannot kill peer at %s: no daemon", randsock_file[i]);
	    msg_buf[sizeof(msg_buf) - 1] = '\0';
	    error_error(msg_buf, 0);
	    killfail = 1;
      }
      else {
	killpeer(kill_socket);
	my_shutdown(kill_socket, SHUT_RDWR);
	(void)close(kill_socket);
      }
    }

    for (i = 0; i < numsock_tcp; i++) {
      if ((kill_socket = tcp_connect(randsock_tcp[i], 0, 1)) < 0) {
	(void)snprintf(msg_buf, sizeof(msg_buf),
		       "cannot kill peer at %s: no daemon", randsock_tcp[i]);
	msg_buf[sizeof(msg_buf) - 1] = '\0';
	error_error(msg_buf, 0);
	/*
	 * Do no exit because there may be more work to do!
	 */
      }
      else {
	killpeer(kill_socket);
	my_shutdown(kill_socket, SHUT_RDWR);
	(void)close(kill_socket);
      }
    }
    /*
     * Everything has been done!
     */
    if (killfail)
      exit(EXIT_FAILURE);
    else
      exit(EXIT_SUCCESS);
  }


  /*
   * Set up the UNIX domain sockets to service entropy requests
   */
  for (i = 0; i < numsock_file; i++) {

    /*
     * Check out what is already available at the place of the future
     * egd-socket. If we cannot reach the place to be with "stat()",
     * do not continue, as something may be seriously wrong. If the
     * entry does already exist, remove it, but only if it is of type
     * socket! If removal fails for whatever reason, stop, as something
     * may be seriously wrong.
     */
    statval = stat(randsock_file[i], &randsock_file_stat);
    if ((statval < 0) && (errno != ENOENT)) {
      (void)snprintf(msg_buf, sizeof(msg_buf), "Cannot stat socket position %s",
      	randsock_file[i]);
      msg_buf[sizeof(msg_buf) - 1] = '\0';
      error_fatal(msg_buf, errno);
      exit(EXIT_FAILURE);
    }
    else if (!statval) {
      if (!IS_UNIX_SOCKET(randsock_file_stat.st_mode)) {
	(void)snprintf(msg_buf, sizeof(msg_buf), "Will not touch %s: no socket",
		randsock_file[i]);
	msg_buf[sizeof(msg_buf) - 1] = '\0';
	error_fatal(msg_buf, errno);
	exit(EXIT_FAILURE);
      } else {
	if ((kill_socket = unix_connect(randsock_file[i], 0, 1)) >= 0) {
	  (void)snprintf(msg_buf, sizeof(msg_buf),
		  "socket %s already used by another process, exiting.\n",
		  randsock_file[i]);
	  msg_buf[sizeof(msg_buf) - 1] = '\0';
	  error_fatal(msg_buf, 0);
	  exit(EXIT_FAILURE);
	}

	/*
         * Now there is a socket without service listening on it: we can
	 * remove it.
	 */
	if (unlink(randsock_file[i]) < 0) {
	  (void)snprintf(msg_buf, sizeof(msg_buf),
	  	"Cannot remove already existing entry %s", randsock_file[i]);
	  msg_buf[sizeof(msg_buf) - 1] = '\0';
	  error_fatal(msg_buf, errno);
	  exit(EXIT_FAILURE);
	}
      }
    }

    service_socket[service_count++] = unix_listen(randsock_file[i], 1, 1);
    if (chmod(randsock_file[i], socketmode) < 0) {
      (void)snprintf(msg_buf, sizeof(msg_buf), "Could not chmod socket to %04o",
		     socketmode);
      msg_buf[sizeof(msg_buf) - 1] = '\0';
      error_fatal(msg_buf, errno);
      exit(EXIT_FAILURE);
    }
    count_openfd++;
  }

  /*
   * Set up the TCP sockets to service entropy requests
   */
  for (i = 0; i < numsock_tcp; i++) {
    service_socket[service_count++] = tcp_listen(randsock_tcp[i], 1, 1);
    count_openfd++;
  }


  /*
   * If not in debugging mode, make myself a daemon 
   */
  if (!debug)
  {
    ret = fg ? fg : daemon(0, 0);
    if (ret < 0)
    {
      error_fatal("Could not daemonize", errno);
      exit(EXIT_FAILURE);
    }

    if ((pwent = getpwuid(getuid())) != NULL)
      username = pwent->pw_name;
    /*
     * The setup is finished, and we are daemonized, so we must now switch
     * to syslog...
     */
    (void)snprintf(msg_buf, sizeof(msg_buf),
    	"prngd %s (%s) started up for user %s", PRNGD_VERSION, PRNGD_DATE,
	username);
    msg_buf[sizeof(msg_buf) - 1] = '\0';
    error_notice(msg_buf, 0);
    count_openfd++;
    (void)snprintf(msg_buf, sizeof(msg_buf),
    	"have %d out of %d filedescriptors open", count_openfd, limit_openfd);
    msg_buf[sizeof(msg_buf) - 1] = '\0';
    error_notice(msg_buf, 0);
  }

  /*
   * Maybe we have changed our pid, let's add some more of it. 
   */
  seed_internal(NULL);

  if (!killmode && pidfile) {
    memset(pidfilebuffer, 0, sizeof(pidfilebuffer));
    snprintf(pidfilebuffer, sizeof(pidfilebuffer), "%12ld\n", (long)getpid());
    pidfilebuffer[sizeof(pidfilebuffer) - 1] = '\0';
    if (write(pidfilefd, pidfilebuffer, strlen(pidfilebuffer)) < 0) {
      error_fatal("Error writing pidfile", errno);
      exit(EXIT_FAILURE);
    }
    close_pidfile();
    count_openfd--;
  }

  /* Never returns */
  main_loop(service_socket, service_count, seed_stat_interval,
	    seed_ext_interval, entropy_source, MAX_GATHERER_BYTES);

  /*
   * We never get here, but want to make the compiler happy...
   */
  return(EXIT_SUCCESS);
}
